pkg/scheduler/framework/frameworkutils.go

```diff
@@ -0,0 +1,372 @@
+			// Ignore any binding that is of the deleting state.
+		case !isTargetClusterPresent || targetCluster.Spec.State == fleetv1beta1.ClusterStateLeave:
+			// Check if the binding is now obsolete, i.e., it is associated with a cluster that is no longer
+			// in normal operations, but is still of an active or creating state.
+			//
+			// Note that this check is solely for the purpose of detecting a situation where bindings are stranded
+			// on a leaving/left cluster; it does not perform any binding association eligibility check for the cluster.
+			dangling = append(dangling, &binding)
+		case binding.Spec.PolicySnapshotName != policy.Name:
+			// The binding is in the creating or active state, but is no longer associated with the latest scheduling policy snapshot.
+			obsolete = append(obsolete, &binding)
+		case binding.Spec.State == fleetv1beta1.BindingStateCreating:
+			// Check if the binding is of the creating state.
+			creating = append(creating, &binding)
+		case binding.Spec.State == fleetv1beta1.BindingStateActive:
+			// Check if the binding is of the active state.
+			active = append(active, &binding)
+		}
+	}
+
+	return active, creating, obsolete, dangling, deleted
+}
+
+// shouldDownscale checks if the scheduler needs to perform some downscaling, and (if so) how many active or creating bindings
+// it should remove.
+func shouldDownscale(policy *fleetv1beta1.ClusterPolicySnapshot, desired, present, obsolete int) (act bool, count int) {
+	if policy.Spec.Policy.PlacementType == fleetv1beta1.PickNPlacementType && desired <= present {
+		// Downscale only applies to CRPs of the Pick N placement type; and it only applies when the number of
+		// clusters requested by the user is less than the number of currently active + creating bindings combined;
+		// or there are the right number of active + creating bindings, yet some obsolete bindings still linger
+		// in the system.
+		if count := present - desired + obsolete; count > 0 {
+			return true, present - desired
+		}
+	}
+	return false, 0
+}
+
+// refreshSchedulingDecisionsFrom returns a list of scheduling decisions, based on existing
+// bindings and (if applicable) currently present decisions in the policy snapshot status.
+func refreshSchedulingDecisionsFrom(policy *fleetv1beta1.ClusterPolicySnapshot, existing ...[]*fleetv1beta1.ClusterResourceBinding) []fleetv1beta1.ClusterDecision {
+	// Pre-allocate arrays.
+	refreshed := make([]fleetv1beta1.ClusterDecision, 0, len(existing))
+
+	// Build new scheduling decisions.
+	for _, bindingSet := range existing {
+		for _, binding := range bindingSet {
+			refreshed = append(refreshed, binding.Spec.ClusterDecision)
+		}
+	}
+
+	// Move some decisions from unbound clusters, if there are still enough room.
+	if diff := maxClusterDecisionCount - len(refreshed); diff > 0 {
+		current := policy.Status.ClusterDecisions
+		for _, decision := range current {
+			if !decision.Selected {
```

Comment:
how are we sure that the cluster here is not part of the already existing bindings? We update the status because we think the status might be out of sync with the binding

Reply:
Yeah, this one is a missight on my end. It is true that we cannot trust any old decisions on policy snapshot. It has been fixed.